iroh-net/src/netcheck/reportgen/hairpin.rs

```diff
@@ -100,37 +85,31 @@ enum Message {
 
 #[derive(Debug)]
 struct Actor {
-    msg_tx: mpsc::Sender<Message>,
-    msg_rx: mpsc::Receiver<Message>,
+    msg_rx: oneshot::Receiver<Message>,
     netcheck: netcheck::Addr,
     reportgen: reportgen::Addr,
 }
 
 impl Actor {
-    fn addr(&self) -> Addr {
-        Addr {
-            sender: self.msg_tx.clone(),
-        }
-    }
-
-    async fn run(&mut self) {
+    async fn run(self) {
         match self.run_inner().await {
             Ok(_) => debug!("hairpin actor finished successfully"),
             Err(err) => error!("Hairpin actor failed: {err:#}"),
         }
     }
 
-    async fn run_inner(&mut self) -> Result<()> {
-        let sock = UdpSocket::bind("0.0.0.0:0")
+    async fn run_inner(self) -> Result<()> {
+        let socket = UdpSocket::bind_v4(0)
```

Comment:
I did also wonder in the past if this would not be better served by keeping this socket around for a long time.  The netcheck actor stays alive for as long as the magicsock is, so storing this socket there and handing a clone to each hairpin could be an improvement.  Especially now we know this is rather expensive on macos maybe that could be worth the additional complexity.  But if we think so perhaps create an issue and do it in a separate PR.

Reply:
> But if we think so perhaps create an issue and do it in a separate PR.

Yes, lets defer this. I agree it would be better, but doing so required more changes to the code than I wanted when I tried to do it.